### ISSUE TITLE
Bug: `no-self-compare`  should not apply to function return values.

### ISSUE THREAD

**Issue by xiaoxiyao** [eslint/eslint#19611]
### Environment

playground

### What parser are you using?

Default (Espree)

### What did you do?

```js
/*eslint no-self-compare: "error"*/
function parseDate(dateStr) {
  return new Date(dateStr);
}

if (parseDate('December 17, 1995 03:24:00') === parseDate('December 17, 1995 03:24:00')) {
  
}
```
[playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tc2VsZi1jb21wYXJlOiBcImVycm9yXCIqL1xuZnVuY3Rpb24gcGFyc2VEYXRlKGRhdGVTdHIpIHtcbiAgcmV0dXJuIG5ldyBEYXRlKGRhdGVTdHIpO1xufVxuXG5pZiAocGFyc2VEYXRlKCdEZWNlbWJlciAxNywgMTk5NSAwMzoyNDowMCcpID09PSBwYXJzZURhdGUoJ0RlY2VtYmVyIDE3LCAxOTk1IDAzOjI0OjAwJykpIHtcbiAgXG59Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)

### What did you expect to happen?

No error

### What actually happened?

Error: Comparing to itself is potentially pointless.  

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tc2VsZi1jb21wYXJlOiBcImVycm9yXCIqL1xuZnVuY3Rpb24gcGFyc2VEYXRlKGRhdGVTdHIpIHtcbiAgcmV0dXJuIG5ldyBEYXRlKGRhdGVTdHIpO1xufVxuXG5pZiAocGFyc2VEYXRlKCdEZWNlbWJlciAxNywgMTk5NSAwMzoyNDowMCcpID09PSBwYXJzZURhdGUoJ0RlY2VtYmVyIDE3LCAxOTk1IDAzOjI0OjAwJykpIHtcbiAgXG59Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @xiaoxiyao, thanks for the issue!

This rule works simply by comparing tokens on both sides of the operator and reports the problem if they're equal. It doesn't account for potential side effects and assumes that functions return the same value when called with the same arguments, so there can indeed be false positives like the one in your example.

However, I think those are edge cases for which `eslint-disable` comments can be used. In my opinion, we should leave the behavior as is, but add a Known Limitations section to the docs, with few examples of false positives. @eslint/eslint-team thoughts?

**Comment by JoshuaKGoldberg** [comment]
Interesting edge case to bring up. A practical example that might occur in the wild is functions that change and return data. https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tc2VsZi1jb21wYXJlOiBcImVycm9yXCIqL1xuZnVuY3Rpb24gaW5jcmVtZW50VW5sZXNzUmVhY2hlZE1heGltdW0oKSB7XG4gIHJldHVybiBNYXRoLm1heChjb3VudGVyICs9IDEsIDEwKTtcbn1cblxuaWYgKGluY3JlbWVudFVubGVzc1JlYWNoZWRNYXhpbXVtKCkgPT09IGluY3JlbWVudFVubGVzc1JlYWNoZWRNYXhpbXVtKCkpIHtcbiAgLy8gLi4uXG59Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19

```js
/*eslint no-self-compare: "error"*/
function incrementUnlessReachedMaximum() {
  return Math.max(counter += 1, 10);
}

if (incrementUnlessReachedMaximum() === incrementUnlessReachedMaximum()) {
  // ...
}
```

I'm in favor of this request, or at least adding an option to consider function calls to be exemptions from the rule reporting. Many codebases are set up so that function calls are not pure. Semantically, I think a function call in a non-pure codebase implies _"this may change something"_.

IMO we definitely wouldn't want to exempt member accesses to try to account for computed getters. Those semantically aren't supposed to change state - even if developers sometimes do wacky things with them.

**Comment by snitin315** [comment]
I aslo agree with https://github.com/eslint/eslint/issues/19611#issuecomment-2786172724,  we can leave the behavior as is and just mention it in the docs. Accepting this as a docs enhancement.
